### PR TITLE
[SPARK-13343] speculative tasks that didn't commit shouldn't be marked as success

### PR DESCRIPTION
**Pull request by hthuynh2**
**Description**
Currently Speculative tasks that didn't commit can show up as success (depending on timing of commit). This is a bit confusing because that task didn't really succeed in the sense it didn't write anything.
I think these tasks should be marked as KILLED or something that is more obvious to the user exactly what happened. it is happened to hit the timing where it got a commit denied exception then it shows up as failed and counts against your task failures. It shouldn't count against task failures since that failure really doesn't matter.
MapReduce handles these situation so perhaps we can look there for a model.

unknown

**How can this issue happen?**
When both attempts of a task finish before the driver sends command to kill one of them, both of them send the status update FINISHED to the driver. The driver calls TaskSchedulerImpl to handle one successful task at a time. When it handles the first successful task, it sends the command to kill the other copy of the task, however, because that task is already finished, the executor will ignore the command. After finishing handling the first attempt, it processes the second one, although all actions on the result of this task are skipped, this copy of the task is still marked as SUCCESS. As a result, even though this issue does not affect the result of the job, it might cause confusing to user because both of them appear to be successful.

**How does this PR fix the issue?**
The simple way to fix this issue is that when taskSetManager handles successful task, it checks if any other attempt succeeded. If this is the case, it will call handleFailedTask with state==KILLED and reason==TaskKilled(“another attempt succeeded”) to handle this task as begin killed.

**How was this patch tested?**
I tested this manually by running applications, that caused the issue before, a few times, and observed that the issue does not happen again. Also, I added a unit test in TaskSetManagerSuite to test that if we call handleSuccessfulTask to handle status update for 2 copies of a task, only the one that is handled first will be mark as SUCCESS

### CODE REVIEW

**Comment by hthuynh2** [comment]
cc @tgravescs

**Comment by tgravescs** [comment]
ok to test

**Comment by tgravescs** [comment]
+1, changes look good to me. 

@squito   see any problems with this approach?

**Comment by tgravescs** [comment]
@hthuynh2  can you fix the description "as success of failures" , this is just a copy of my typo in the jira.  Can you just change to be "as success"

**Comment by hthuynh2** [comment]
I updated it. Thanks.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala [hunk 1]
```diff
@@ -1371,4 +1371,64 @@ class TaskSetManagerSuite extends SparkFunSuite with LocalSparkContext with Logg
...
+    val accumUpdatesByTask: Array[Seq[AccumulatorV2[_, _]]] = taskSet.tasks.map { task =>
+      task.metrics.internalAccums
+    }
+
+    // Offer resources for 4 tasks to start
+    for ((k, v) <- List(
+      "exec1" -> "host1",
+      "exec1" -> "host1",
+      "exec2" -> "host2",
+      "exec2" -> "host2")) {
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala]: nit: double indent the contents of the `List`

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala [hunk 2]
```diff
@@ -1371,4 +1371,64 @@ class TaskSetManagerSuite extends SparkFunSuite with LocalSparkContext with Logg
...
+    assert(task5.taskId === 4)
+    assert(task5.executorId === "exec1")
+    assert(task5.attemptNumber === 1)
+    sched.backend = mock(classOf[SchedulerBackend])
+
+    // Complete one attempt for the running task
+    manager.handleSuccessfulTask(3, createTaskResult(3, accumUpdatesByTask(3)))
+    // Verify that it kills other running attempt
+    verify(sched.backend).killTask(4, "exec1", true, "another attempt succeeded")
+    // Complete another attempt for the running task
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala]: can you expand this comment to explain why you're doing this?  without looking at the bug, it's easy to think this part is wrong, but in fact its the most important part of your test.  eg:

There is a race between the scheduler asking to kill the other task, and that task actually finishing.  We simulate what happens if the other task finishes before we kill it.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala [hunk 3]
```diff
@@ -1371,4 +1371,64 @@ class TaskSetManagerSuite extends SparkFunSuite with LocalSparkContext with Logg
...
+    manager.handleSuccessfulTask(4, createTaskResult(3, accumUpdatesByTask(3)))
+
+    assert(manager.taskInfos(3).successful == true)
+    assert(manager.taskInfos(4).killed == true)
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala]: it seems the main thing you're trying to change here is what gets passed to `DAGScheduler.taskEnded`, so shouldn't you be verifying that here?

**Review by squito — COMMENTED** [review]
approach makes sense to me, I have some suggestions for making the test a bit better.

**Comment by jiangxb1987** [comment]
IIUC this speculative task is not really killed right ? It is actually ignored. Does that worth it to add a new TaskState for this case ?

**Comment by hthuynh2** [comment]
@jiangxb1987 yes, you are correct that it is actually ignored. I think it doesn't worth to add a new TaskState because we might need to add changes in many places but does not add much benefit. Instead, I think we can add some message to the kill message to differentiate it from task that is actually killed and to inform the user.

**Comment by hthuynh2** [comment]
@squito Thanks for the suggestions. I updated it. Could you please have a look at it to see if there is anything else I need to change? Thanks.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 4]
```diff
@@ -723,6 +723,13 @@ private[spark] class TaskSetManager(
   def handleSuccessfulTask(tid: Long, result: DirectTaskResult[_]): Unit = {
     val info = taskInfos(tid)
     val index = info.index
+    // Check if any other attempt succeeded before this and this attempt has not been handled
+    if (successful(index) && killedByOtherAttempt(index)) {
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: For completeness, we will also need to 'undo' the changes in `enqueueSuccessfulTask` : to reverse the counters in `canFetchMoreResults`.


(Orthogonal to this PR): Looking at use of `killedByOtherAttempt`, I see that there is a bug in `executorLost` w.r.t how it is updated - hopefully a fix for SPARK-24755 wont cause issues here.

**Comment by tgravescs** [comment]
lets review https://github.com/apache/spark/pull/21729 before this since its changing the type on killedByOtherAttempt

**Comment by tgravescs** [comment]
#21729 has been merged @hthuynh2  can you update this one?

**Comment by hthuynh2** [comment]
@tgravescs I updated it. Can you please have a look at it when you have time. Thank you.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 5]
```diff
@@ -723,6 +723,21 @@ private[spark] class TaskSetManager(
...
     val info = taskInfos(tid)
     val index = info.index
+    // Check if any other attempt succeeded before this and this attempt has not been handled
+    if (successful(index) && killedByOtherAttempt.contains(tid)) {
+      calculatedTasks -= 1
+
+      val resultSizeAcc = result.accumUpdates.find(a =>
+        a.name == Some(InternalAccumulator.RESULT_SIZE))
+      if (resultSizeAcc.isDefined) {
+        totalResultSize -= resultSizeAcc.get.asInstanceOf[LongAccumulator].value
```
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: the downside here is we already incremented and other tasks could have checked and failed before we decrement, but unless someone else has a better idea this is better then it is now.

**Review by tgravescs — CHANGES_REQUESTED** [review]

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala (hunk 5 above)
Comment (mridulm) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: I agree, I dont see a better option.

**Review by mridulm — COMMENTED** [review]
LGTM, pending @tgravescs's suggestions.

**Review by jiangxb1987 — APPROVED** [review]
LGTM

**Comment by hthuynh2** [comment]
@tgravescs Can you please run the test again, thank you.

**Comment by tgravescs** [comment]
test this please

**Comment by HyukjinKwon** [comment]
test this please

**Comment by tgravescs** [comment]
test this please

**Comment by tgravescs** [comment]
test this please

**Comment by tgravescs** [comment]
test this please

**Comment by tgravescs** [comment]
test this please

**Comment by squito** [comment]
I kicked off the test manually at https://spark-prs.appspot.com/users/hthuynh2.  I dunno why the test triggering via comments stops workign on some prs

**Comment by tgravescs** [comment]
merged to master, thanks @hthuynh2

**Pull request closed without merging**